engines/sci/graphics/gfxdrivers.cpp

```diff
@@ -0,0 +1,508 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "common/events.h"
+#include "common/system.h"
+
+#include "graphics/cursorman.h"
+#include "graphics/paletteman.h"
+
+#include "sci/graphics/gfxdrivers.h"
+#include "sci/resource/resource.h"
+#include "sci/sci.h"
+#include "sci/version.h"
+
+namespace Sci {
+
+Common::Point GfxDriver::getMousePos() const {
+	return g_system->getEventManager()->getMousePos();
+}
+
+GfxDefaultDriver::GfxDefaultDriver(uint16 screenWidth, uint16 screenHeight) : GfxDriver(screenWidth, screenHeight, 0, 1) {
+	switch (g_sci->getResMan()->getViewType()) {
+	case kViewEga:
+		_numColors = 16;	// QFG PC-98 with 8 colors also reports 16 here
+		break;
+	case kViewAmiga:
+		_numColors = 32;
+		break;
+	case kViewAmiga64:
+		_numColors = 64;
+		break;
+	case kViewVga:
+	case kViewVga11:
+		_numColors = 256;
+		break;
+	default:
+		break;
+	}
+
+	if (_numColors == 0)
+		error("GfxDefaultDriver: Unknown view type");
+}
+
+void GfxDefaultDriver::setPalette(const byte *colors, uint start, uint num) {
+	g_system->getPaletteManager()->setPalette(colors, start, num);
+}
+
+void GfxDefaultDriver::copyRectToScreen(const byte *src, int pitch, int x, int y, int w, int h) {
+	g_system->copyRectToScreen(src, pitch, x, y, w, h);
+}
+
+void GfxDefaultDriver::replaceCursor(const void *cursor, uint w, uint h, int hotspotX, int hotspotY, uint32 keycolor) {
+	CursorMan.replaceCursor(cursor, w, h, hotspotX, hotspotY, keycolor);
+}
+
+SCI0_DOSPreVGADriver::SCI0_DOSPreVGADriver(int numColors, int screenW, int screenH, int horizontalAlignment) : GfxDriver(screenW, screenH, numColors, horizontalAlignment), _palNeedUpdate(true), _colors(nullptr), _compositeBuffer(nullptr) {
+	_compositeBuffer = new byte[screenW * screenH]();
+}
+
+SCI0_DOSPreVGADriver::~SCI0_DOSPreVGADriver() {
+	delete[] _compositeBuffer;
+}
+
+bool SCI0_DOSPreVGADriver::checkDriver(const char *const *driverNames, int listSize) {
+	Common::String missing;
+	while (listSize-- && *driverNames) {
+		if (Common::File::exists(*driverNames))
+			return true;
+		if (!missing.empty())
+			missing += " or ";
+		missing += "'" + Common::String(*driverNames) + "'";
+		++driverNames;
+	}
+	warning("Driver file %s not found. Starting game in EGA mode", missing.c_str());
+	return false;
+}
+
+void SCI0_DOSPreVGADriver::assignPalette(const byte *colors) {
+	_colors = colors;
+}
+
+void SCI0_DOSPreVGADriver::setPalette(const byte*, uint, uint) {
+	if (!_palNeedUpdate || !_colors)
+		return;
+	_palNeedUpdate = false;
+	byte *tmp = new byte[768]();
+	memcpy(tmp, _colors, _numColors * 3);
+	g_system->getPaletteManager()->setPalette(tmp, 0, 256);
```

Comment:
Is it important to clear the unused palette entries, or could `_colors` be passed directly here?

Reply:
Just from looking at it, I don't think it is important. We do clear the palette whenever we restart the engine (e. g. after returning to the launcher), so we shouldn't have weird colors left. Also, I don't use any colors outside the valid range (for some engines it make sense to have some extra colors for mouse cursor transparency, font shading etc, but not here), so even if there were anyhing left it shouldn't matter.